### PR TITLE
retrive metrics for an app over an interval

### PR DESCRIPTION
**Pull request by boyang9527**
[#127841745]

- add another endpoint to get metrics history
- add hanlder to retrieve metrics from database with start and end time
- change db to return metrics ordered by timestamp
- test cases for server, handler, db  and main

### CODE REVIEW

File: src/metricscollector/cmd/metricscollector/main.go [hunk 1]
```diff
@@ -60,13 +62,18 @@ func main() {
...
 	logger.Info("create-noaa-client", map[string]interface{}{"dopplerUrl": dopplerUrl})
 	tlsConfig := &tls.Config{InsecureSkipVerify: true}
 	noaa := consumer.New(dopplerUrl, tlsConfig, nil)
 	noaa.RefreshTokenFrom(cfClient)
 
-	httpServer := server.NewServer(logger, conf.Server, cfClient, noaa)
+	var database db.DB
+	database, err = sqldb.NewSQLDB(&conf.Db, logger.Session("db"))
+	if err != nil {
+		logger.Error("failed to connect database", err, lager.Data{"conf": conf.Db})
```
Comment (fraenkel) [inline src/metricscollector/cmd/metricscollector/main.go]: Shouldn't we exit?

File: src/metricscollector/server/memory_metric_handler.go [hunk 2]
```diff
@@ -44,20 +48,20 @@ func (h *MemoryMetricHandler) GetMemoryMetric(w http.ResponseWriter, r *http.Req
...
 
-	h.logger.Debug("get-container-metrics-from-doppler", lager.Data{"container-metrics": containerMetrics})
 	metric := metrics.GetMemoryMetricFromContainerMetrics(appId, containerMetrics)
-
-	body, err := json.Marshal(metric)
+	var body []byte
+	body, err = json.Marshal(metric)
 	if err != nil {
-		h.logger.Error("marshal-memory-metrics", err, lager.Data{"metric": metric})
+		h.logger.Error("Get-memory-metrics-marshal", err, lager.Data{"metric": metric})
```
Comment (fraenkel) [inline src/metricscollector/server/memory_metric_handler.go]: would it be useful to know the appId?

File: src/metricscollector/server/memory_metric_handler.go [hunk 3]
```diff
@@ -67,3 +71,75 @@ func (h *MemoryMetricHandler) GetMemoryMetric(w http.ResponseWriter, r *http.Req
...
+	appId := vars["appid"]
+	startParam := r.URL.Query()["start"]
+	endParam := r.URL.Query()["end"]
+	h.logger.Debug("get-memeory-metric-history", lager.Data{"appId": appId, "start": startParam, "end": endParam})
+
+	var err error
+	start := int64(0)
+	end := int64(-1)
+
+	if startParam != nil {
```
Comment (fraenkel) [inline src/metricscollector/server/memory_metric_handler.go]: not needed.
len(nil) == 0

File: src/metricscollector/server/memory_metric_handler_test.go [hunk 4]
```diff
@@ -106,4 +115,194 @@ var _ = Describe("MemoryMetricHandler", func() {
...
+					req, err = http.NewRequest(http.MethodGet, "http://localhost/v1/apps/an-app-id/metrics_history/memory?start=123&start=231", nil)
+					Expect(err).ToNot(HaveOccurred())
+
+				})
+
+				It("returns 400", func() {
+					Expect(resp.Code).To(Equal(http.StatusBadRequest))
+
+					errJson := &ErrorResponse{}
+					d := json.NewDecoder(resp.Body)
```
Comment (fraenkel) [inline src/metricscollector/server/memory_metric_handler_test.go]: Might as well use json.Unmarshal() to reduce lines.

File: src/metricscollector/server/memory_metric_handler_test.go [hunk 5]
```diff
@@ -106,4 +115,194 @@ var _ = Describe("MemoryMetricHandler", func() {
...
+					err = d.Decode(errJson)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(errJson.Code).To(Equal("Bad-Request"))
```
Comment (fraenkel) [inline src/metricscollector/server/memory_metric_handler_test.go]: Why not just do a Equal(&ErrorResponse{
Code: "Bad-Request",
Message: ".....",
})

File: src/metricscollector/server/memory_metric_handler_test.go [hunk 6]
```diff
@@ -106,4 +115,194 @@ var _ = Describe("MemoryMetricHandler", func() {
...
+					Expect(errJson.Message).To(Equal("Error parsing end time"))
+				})
+			})
+
+		})
+
+		Context("when request query string is valid", func() {
```
Comment (fraenkel) [inline src/metricscollector/server/memory_metric_handler_test.go]: If you reordered your tests and got the base cases out of the way, you need to check less later on.
While it is good practice to make sure everything is valid, you can also get away with just checking for the difference/new behavior.

File: src/metricscollector/cmd/metricscollector/main.go (hunk 1 above)
Comment (boyang9527) [inline src/metricscollector/cmd/metricscollector/main.go]: added exit and close the database

File: src/metricscollector/server/memory_metric_handler.go (hunk 2 above)
Comment (boyang9527) [inline src/metricscollector/server/memory_metric_handler.go]: added in log data

File: src/metricscollector/server/memory_metric_handler.go (hunk 3 above)
Comment (boyang9527) [inline src/metricscollector/server/memory_metric_handler.go]: still need to check whether "startParam == nil" or "len(startParam) == 0". In this case, start will remain the default value 0. 
Anyway changed to use len() for all cases

File: src/metricscollector/server/memory_metric_handler_test.go (hunk 4 above)
Comment (boyang9527) [inline src/metricscollector/server/memory_metric_handler_test.go]: all changed

File: src/metricscollector/server/memory_metric_handler_test.go (hunk 5 above)
Comment (boyang9527) [inline src/metricscollector/server/memory_metric_handler_test.go]: all changed.

File: src/metricscollector/server/memory_metric_handler_test.go (hunk 6 above)
Comment (boyang9527) [inline src/metricscollector/server/memory_metric_handler_test.go]: changed the order to reduce the checks.

**Pull request merged**